File: Level 3/외벽 점검/solution.py
from itertools import permutations


def solution(n, weak, dist):
    weak_length = len(weak)

    for i in range(weak_length):
        weak.append(weak[i] + n)

    answer = len(dist) + 1

    for i in range(weak_length):
        for friends in permutations(dist, len(dist)):
            count = 1
            position = weak[i] + friends[count - 1]
            for j in range(i, i + weak_length):
                if position < weak[j]:
                    count += 1
                    if count > len(dist):
                        break
                    position = weak[j] + friends[count - 1]
            answer = min(answer, count)

    return answer if answer <= len(dist) else -1

File: Level 3/외벽 점검/solution.js
function solution(n, weak, dist) {
  const weakLength = weak.length;
  for (let i = 0; i < weakLength; i++) {
    weak.push(weak[i] + n);
  }

  let answer = dist.length + 1;

  for (let i = 0; i < weakLength; i++) {
    for (const friends of getPermutations(dist, dist.length)) {
      let count = 1;
      let position = weak[i] + friends[count - 1];
      for (let j = i; j < i + weakLength; j++) {
        if (position < weak[j]) {
          count++;
          if (count > dist.length) {
            break;
          }
          position = weak[j] + friends[count - 1];
        }
      }
      answer = Math.min(answer, count);
    }
  }

  return answer <= dist.length ? answer : -1;
}

function getPermutations(arr, n) {
  if (n === 0) {
    return [[]];
  }

  const result = [];

  arr.forEach((fixed, index) => {
    const rest = [...arr];
    rest.splice(index, 1);
    const permutations = getPermutations(rest, n - 1);
    const combine = permutations.map((permutation) => [fixed, ...permutation]);
    result.push(...combine);
  });

  return result;
}
